include/ginkgo/core/base/reflection.hpp

```diff
@@ -0,0 +1,167 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright (c) 2017-2019, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_CORE_BASE_REFLECTION_HPP_
+#define GKO_CORE_BASE_REFLECTION_HPP_
+
+
+#include <vector>
+
+
+#include <ginkgo/core/base/lin_op.hpp>
+
+
+namespace gko {
+
+
+/**
+ * The Reflection class can be used to be as (I + coef * U * V)
+ *
+ * @tparam ValueType  precision of input and result vectors
+ *
+ * @ingroup LinOp
+ */
+template <typename ValueType = default_precision>
+class Reflection : public EnableLinOp<Reflection<ValueType>>,
+                   public EnableCreateMethod<Reflection<ValueType>> {
+    friend class EnablePolymorphicObject<Reflection, LinOp>;
+    friend class EnableCreateMethod<Reflection>;
+
+public:
+    using value_type = ValueType;
+
+    /**
+     * Returns the operator U of the reflection.
+     *
+     * @return the operator U
+     */
+    const std::shared_ptr<const LinOp> get_u_operator() const noexcept
```

Comment:
This does not seems descriptive enough to me. It ties reflection to two variables, which I don't think is intuitive. So in general this matrix is called [rank-one perturbation](https://math.stackexchange.com/questions/791530/rank-one-perturbation-proof). But I think the operator does not have a proper name, but I think it can be called a rank one perturbation operator.

We should maybe discuss the names for these getter function and also maybe the names of the class itself.